### PR TITLE
[RFC][DNM] drivers: adc: new API proposal

### PR DESCRIPTION
**Pull request by anangl**
This is an initial proposal of a new API for the ADC driver, so that we could start discussion and conclude what is actually needed in the API, and what will be convenient to use.
I tried to keep similarity with the new SPI API (#5921).

A few questions/doubts that I have:
1. Is there a need to use a bigger type than `s16_t` for sample values?
2. How can gain/prescaling settings be effectively passed to the driver? I suppose that these settings will vary significantly between different controllers.
3. Are there any other important configuration settings (of the ADC as a whole, or particular channels) that I haven't included yet?

@nashif @carlescufi @tbursztyka @MaureenHelm @andrewboie @galak

### CODE REVIEW

File: include/adc.h [hunk 1]
```diff
@@ -27,119 +27,152 @@ extern "C" {
...
- *
- * This structure defines a list of sequence entries
- * used to execute a sequence of samplings.
+ * @brief ADC sampling definition structure.
  */
-struct adc_seq_table {
-	/* Pointer to a sequence entry array. */
-	struct adc_seq_entry *entries;
+struct adc_channel_cfg {
+	u8_t input_positive;
```
Comment (tbursztyka) [inline include/adc.h]: some description?

File: include/adc.h [hunk 2]
```diff
@@ -27,119 +27,152 @@ extern "C" {
...
+
+	/** ADC resolution.
+	 * For single-ended channels the sample values are from range:
+	 *   0 .. 2^resolution - 1,
+	 * for differential ones:
+	 *   -2^resolution .. 2^resolution - 1. */
+	u8_t resolution;
+	/** Oversampling setting.
+	 *  Each sample is averaged from 2^oversampling results. */
+	u8_t oversampling;
```
Comment (tbursztyka) [inline include/adc.h]: so 0 would mean bare result?

File: include/adc.h [hunk 3]
```diff
@@ -27,119 +27,152 @@ extern "C" {
...
-struct adc_seq_entry {
-	/** Clock ticks delay before sampling the ADC. */
-	s32_t sampling_delay;
-
-	/** Buffer pointer where the sample is written.*/
-	u8_t *buffer;
+struct adc_sampling {
+	/** Clock ticks delay before starting the sampling (counted from the
+	 *  start of the previous sampling in the sequence, if any). */
+	u32_t delay;
```
Comment (tbursztyka) [inline include/adc.h]: ticks is hw dependent. Wouldn't it be better to work on µs or ms? (Well, µs is only supported through busy wait loop)

**Review by tbursztyka — COMMENTED** [review]
samples value size will depend on resolution, so 8, 16, 24, 32, 64...?
Instead, I would make in void *, and depending on the resolution, the driver would fill it in accordingly.
Only thing would be that driver would check the buffer length vs resolution at first.
Just thinking out loud, I don't have much experience in ADCs (besides the awful adc_ti_adc108s102 I did back in the days)

File: include/adc.h [hunk 4]
```diff
@@ -27,119 +27,152 @@ extern "C" {
...
 
 /**
- * @brief Set a read request.
- *
- * This routine sends a read or sampling request to the ADC hardware block.
- * A sequence table describes the read request.
- * The routine returns once the ADC completes the read sequence.
- * The sample data can be retrieved from the memory buffers in
- * the sequence table structure.
+ * @brief Set an asynchronous read request.
```
Comment (galak) [inline include/adc.h]: This comment seems incorrect.

Comment (galak) [reply include/adc.h]: We should document the following - is it required to call adc_channel_setup before adc_read*?

**Review by galak — CHANGES_REQUESTED** [review]
I get why adc_channel_setup() takes a dev pointer and the others dont, but the inconsistency is unfortunate.

File: include/adc.h [hunk 5]
```diff
@@ -27,119 +27,152 @@ extern "C" {
...
-
-static inline int _impl_adc_read(struct device *dev,
-				 struct adc_seq_table *seq_table)
+__syscall int adc_channel_setup(struct device *dev,
+				u8_t channel_id,
+				const struct adc_channel_cfg *channel_cfg);
+
+static inline
+int _impl_adc_channel_setup(struct device *dev,
+			    u8_t channel_id,
```
Comment (galak) [inline include/adc.h]: any reason channel_id isn't in adc_channel_cfg?

File: include/adc.h [hunk 6]
```diff
@@ -27,119 +27,152 @@ extern "C" {
...
  */
-struct adc_seq_table {
-	/* Pointer to a sequence entry array. */
-	struct adc_seq_entry *entries;
+struct adc_channel_cfg {
+	u8_t input_positive;
+	u8_t input_negative;
+	/**
+	 *  A bit field with the following parts:
+	 *  [5 :  0] - acquisition time, in microseconds
```
Comment (vaussard) [inline include/adc.h]: 6 bits for the acquisition time looks a bit short, as 63 microseconds may be too small if you sample a high-impedance source or your ADC runs on a very slow clock.
On the other hand, 1us is rather high if you go high speed. For the STM32L4A6, I can setup Tacq as low as 30 nanoseconds.
So maybe we need a more flexible approach to specify the conversion time?

File: include/adc.h [hunk 7]
```diff
@@ -27,119 +27,152 @@ extern "C" {
...
+struct adc_sampling {
+	/** Clock ticks delay before starting the sampling (counted from the
+	 *  start of the previous sampling in the sequence, if any). */
+	u32_t delay;
 
-	/** Length of the sampling buffer.*/
-	u32_t buffer_length;
+	/** Pointer to a buffer where the samples are to be written.
+	 *  Each sample value is sign-extended to 16 bits. */
+	s16_t *buffer;
```
Comment (vaussard) [inline include/adc.h]: s16_t is not enough, I used external 24-bit ADC in the past. And you need more bits when doing oversampling.

**Comment by vaussard** [comment]
A missing feature is the ability to setup the trigger source. Here a software trigger is implied, but we have to use hardware triggers sometime. I do not see a straightforward way to setup it as it is usually hardware dependant, but maybe some kind of driver-specific API (like uart_drv_cmd() or Linux ioctl) ?

**Comment by anangl** [comment]
I will be on vacation all next week. I'll address all the comments (already existing ones and any new that may appear, and are welcome) when I'm back.

File: include/adc.h [hunk 8]
```diff
@@ -27,119 +27,152 @@ extern "C" {
...
- * to define a sample from a specific channel.
+ * @brief ADC sampling definition structure.
  */
-struct adc_seq_entry {
-	/** Clock ticks delay before sampling the ADC. */
-	s32_t sampling_delay;
-
-	/** Buffer pointer where the sample is written.*/
-	u8_t *buffer;
+struct adc_sampling {
```
[the remark this answers is not in the document]
Comment (MaureenHelm) [reply include/adc.h]: It would be more consistent with spi to separate them

File: include/adc.h (hunk 3 above)
Comment (MaureenHelm) [reply include/adc.h]: I'm torn on this one. It's a pain to do the conversion in the driver and often inefficient due to divide operations, but I can see the value in having common units like us or ms for the application.

File: include/adc.h (hunk 7 above)
Comment (MaureenHelm) [reply include/adc.h]: Agreed

File: include/adc.h [hunk 9]
```diff
@@ -27,119 +27,152 @@ extern "C" {
...
+struct adc_channel_cfg {
+	u8_t input_positive;
+	u8_t input_negative;
+	/**
+	 *  A bit field with the following parts:
+	 *  [5 :  0] - acquisition time, in microseconds
+	 *  [6]      - mode, 0: single-ended (only positive input is specified)
+	 *                   1: differential (both inputs need to be specified)
+	 *  [7 :  ?] - gain/prescaling setting
+	 *  anything else ?
```
[the remark this answers is not in the document]
Comment (MaureenHelm) [reply include/adc.h]: Do these things ever change on a per-channel basis? They don't on NXP devices

File: include/adc.h [hunk 10]
```diff
@@ -27,119 +27,152 @@ extern "C" {
...
 struct adc_driver_api {
-	/** Pointer to the enable routine. */
-	void (*enable)(struct device *dev);
-
-	/** Pointer to the disable routine. */
-	void (*disable)(struct device *dev);
-
-	/** Pointer to the read routine. */
-	int (*read)(struct device *dev, struct adc_seq_table *seq_table);
+	int (*read)(const struct adc_config *config,
```
Comment (MaureenHelm) [inline include/adc.h]: Please add the dev pointer argument back and remove it from the adc_config struct

**Review by MaureenHelm — COMMENTED** [review]
> A missing feature is the ability to setup the trigger source. Here a software trigger is implied, but we have to use hardware triggers sometime. I do not see a straightforward way to setup it as it is usually hardware dependant,

Agreed

**Comment by anangl** [comment]
Tried to address gathered feedback, to some extent at least.
Please consider it still as a work in progress, more a base for discussion than a real code.

File: include/adc.h [hunk 11]
```diff
@@ -27,119 +27,225 @@ extern "C" {
...
+			     const struct adc_sequence *sequence,
+			     struct k_poll_signal *async);
 
-static inline void _impl_adc_disable(struct device *dev)
+static inline int _impl_adc_read_async(struct device *dev,
+				       const struct adc_sequence *sequence,
+				       struct k_poll_signal *async)
 {
-	const struct adc_driver_api *api = dev->driver_api;
+	const struct adc_driver_api *api = config->dev->driver_api;
```
Comment (tbursztyka) [inline include/adc.h]: tiny left over from previous PR push

Comment (anangl) [reply include/adc.h]: Right. I haven't tried asynchronous part yet. That's why I haven't got any compile error for this.

File: include/adc.h [hunk 12]
```diff
@@ -27,119 +27,225 @@ extern "C" {
...
 struct adc_driver_api {
-	/** Pointer to the enable routine. */
-	void (*enable)(struct device *dev);
-
-	/** Pointer to the disable routine. */
-	void (*disable)(struct device *dev);
-
-	/** Pointer to the read routine. */
-	int (*read)(struct device *dev, struct adc_seq_table *seq_table);
+	int (*read)(struct device *dev,
```
Comment (tbursztyka) [inline include/adc.h]: to be consistent  with other APIs, create function typedefs for the functions first

File: include/adc.h [hunk 13]
```diff
@@ -27,119 +27,225 @@ extern "C" {
...
+	 *  -2^(resolution-1) .. 2^(resolution-1) - 1.
+	 */
+	u8_t resolution;
+
+	/**
+	 * Oversampling setting.
+	 * Each sample is averaged from 2^oversampling results.
+	 */
+	u8_t oversampling;
+
```
Comment (tbursztyka) [inline include/adc.h]: I would reorder the attributes so these 2 u8_t would be at the end of the struct.

File: include/adc.h [hunk 14]
```diff
@@ -27,119 +27,225 @@ extern "C" {
...
+/* Forward declaration of the adc_sequence structure. */
+struct adc_sequence;
+
+typedef int (*adc_sequence_callback_t)(struct device *dev,
+				       const struct adc_sequence *sequence,
+				       u8_t sampling_number);
+
+struct adc_sequence_continuation {
+	/** Number of consecutive samplings to perform. */
+	u8_t sampling_count;
```
Comment (tbursztyka) [inline include/adc.h]: I would move this attribute at the end of the struct

File: include/adc.h [hunk 15]
```diff
@@ -27,119 +27,225 @@ extern "C" {
...
+	 * Use the ADC_ACQ_TIME macro to compose the value for this field.
+	 * Particular drivers do not necessarily support all the possible units.
+	 * Value range is 0..16383 for a given unit.
+	 */
+	u16_t acquisition_time;
+
+	/** ADC reference selection. */
+	enum adc_reference  reference;
+	/** Channel voltage prescaling selection. */
+	enum adc_prescaling prescaling;
```
Comment (tbursztyka) [inline include/adc.h]: move these 2 attributes at the beginning of the struct

**Review by tbursztyka — COMMENTED** [review]
tiny changes requested. (mostly on struct alignment)

File: include/adc.h [hunk 16]
```diff
@@ -27,119 +27,225 @@ extern "C" {
...
+	 * Pointer to a buffer where the samples are to be written.
+	 * It must be of proper size, taking into account the number of samples
+	 * and used ADC resolution.
+	 */
+	void *buffer;
 
-	/** Buffer pointer where the sample is written.*/
-	u8_t *buffer;
+	/**
+	 * Channels to be used for getting consecutive samples.
```
Comment (MaureenHelm) [inline include/adc.h]: Do they need to be unique? (no duplicates)

File: include/adc.h [hunk 17]
```diff
@@ -27,119 +27,225 @@ extern "C" {
...
+	 */
+	const u8_t *channel_ids;
 
-	/** Length of the sampling buffer.*/
-	u32_t buffer_length;
+	/**
+	 * Number of samples to read in the sampling.
+	 */
+	u8_t sample_count;
```
Comment (MaureenHelm) [inline include/adc.h]: Maybe this should be channel_count? It's a little confusing to have "samplings" and "samples" be different things

File: include/adc.h [hunk 18]
```diff
@@ -27,119 +27,225 @@ extern "C" {
...
+};
 
-	/** Channel ID that should be sampled from the ADC */
-	u8_t channel_id;
+enum adc_trigger
+{
+	ADC_TRIGGER_DELAY_TICKS,
+	ADC_TRIGGER_DELAY_US,
+	ADC_TRIGGER_DELAY_MS,
```
Comment (MaureenHelm) [inline include/adc.h]: Are the triggers one-off or recurring (such as from a hardware timer)? Can you explain the difference between the `DELAY` triggers and the `SAMPLE_RATE` trigger?

File: include/adc.h [hunk 19]
```diff
@@ -27,119 +27,225 @@ extern "C" {
...
+
+typedef int (*adc_sequence_callback_t)(struct device *dev,
+				       const struct adc_sequence *sequence,
+				       u8_t sampling_number);
+
+struct adc_sequence_continuation {
+	/** Number of consecutive samplings to perform. */
+	u8_t sampling_count;
+
+	/** Type of trigger that starts each sampling. */
```
Comment (MaureenHelm) [inline include/adc.h]: A sampling is a group of channels?

File: include/adc.h [hunk 20]
```diff
@@ -27,119 +27,225 @@ extern "C" {
...
+	 * Pointer to structure defining the way in which the sequence is to be
+	 * continued. If NULL, the sequence consists of a single sampling.
+	 */
+	const struct adc_sequence_continuation *continuation;
+};
+
+enum adc_prescaling
+{
+	ADC_PRESCALING_1_6,
+	ADC_PRESCALING_1_5,
```
Comment (MaureenHelm) [inline include/adc.h]: Are these numbers non-integer gain factors? We will need more values in the enum, as the nxp adc supports gain factors of {1,2,4,8,16,32,64}

File: include/adc.h (hunk 17 above)
Comment (anangl) [reply include/adc.h]: I simplified the structure and used a bit-mask for specifying the channels to be included in the sampling. The previous proposal with the `channel_ids` pointer and `sample_count` was too confusing and probably troublesome in implementation (regarding duplicates, as you pointed out in the comment above, and order of the channels).

My idea was to define a sampling as reading of samples from one or more channels. Such samplings can be then arranged in sequences with optional triggers. I added more comments in the header file and changed some names to hopefully present the idea clearer. I am open to change the terminology to a better one. English is not my native language so the names I propose may not be the best (or even good enough) ones.

File: include/adc.h (hunk 20 above)
Comment (anangl) [reply include/adc.h]: Yes, these are fractional factors. I added proper comments, and renamed "prescaling" to "gain".

File: include/adc.h (hunk 19 above)
Comment (anangl) [reply include/adc.h]: Yes. I added better, hopefully, descriptions of the structure and its fields.

**Comment by anangl** [comment]
Updated the proposal according to the recent suggestions. Added more descriptions to make things clearer.

**Comment by anangl** [comment]
I updated the proposal once more. I got rid of the `adc_sampling` structure - its fields became members of the `adc_sequence` structure. I think this way it will be less confusing and simpler for example for single channel samplings, and the possibility of having a sequence of samplings each performed for a different sets of selected channels, which is no longer present, was actually not that much usable.
I added also actions that the application can pass from the callback to driver to alter the performed sequence.
And finally, I updated the adc_api test to show my vision of the usage of the proposed API.

**Comment by tbursztyka** [comment]
Seems fine to me.
You seem to have forgotten the drivers update? (to get your test running)

**Comment by anangl** [comment]
Replaced by #7691.

**Pull request closed without merging**